text/3874-build-std-always.md

```diff
@@ -0,0 +1,1538 @@
+
+Furthermore, custom targets will be destabilised in rustc (as in [rust#71009]).
+This will not be a significant breaking change as custom targets cannot
+effectively be used currently without nightly (needing build-std to have
+`core`).
+
+Custom targets can still be used with build-std on nightly toolchains provided
+that `-Zunstable-options` is provided to Cargo.
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why disallow custom targets?*][rationale-disallow-custom-targets]
+
+*See the following sections for future possibilities:*
+
+- [*Allow custom targets with build-std*][future-custom-targets]
+
+## Preventing implicit sysroot dependencies
+[preventing-implicit-sysroot-dependencies]: #preventing-implicit-sysroot-dependencies
+
+Cargo will pass a new flag to rustc which will prevent rustc from loading
+top-level dependencies from the sysroot ([?][rationale-root-sysroot-deps]).
+
+> [!NOTE]
+>
+> rustc could add a `--no-implicit-sysroot-deps` flag with this behaviour. For
+> example, writing `extern crate foo` in a crate will not load `foo.rlib` from
+> the sysroot if it is present, but if an `--extern noprelude:bar.rlib` is
+> provided which depends on a crate `foo`, rustc will look in
+> `-L dependency=...` paths and the sysroot for it.
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why prevent rustc from loading root dependencies from the sysroot?*][rationale-root-sysroot-deps]
+
+## Vendored `rust-src`
+[vendored-rust-src]: #vendored-rust-src
+
+When it is necessary to build the standard library, Cargo will look for sources
+in a fixed location in the sysroot ([?][rationale-custom-src-path]):
+`lib/rustlib/src`. rustup's `rust-src` component downloads standard library
+sources to this location and will be made a default component. If the sources
+are not found, Cargo will emit an error and recommend the user download
+`rust-src` if using rustup.
+
+`rust-src` will contain the sources for the standard library crates as well as
+its vendored dependencies ([?][rationale-vendoring]). As a consequence sources
+of standard library dependencies will not need be fetched from crates.io.
+
+> [!NOTE]
+>
+> Cargo will not perform any checks to ensure that the sources in `rust-src`
+> have been modified ([?][rationale-src-modifications]). It will be documented
+> that modifying these sources is not supported.
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why not allow the source path for the standard library be customised?*][rationale-custom-src-path]
+- [*Why vendor standard library dependencies?*][rationale-vendoring]
+- [*Why not check if `rust-src` has been modified?*][rationale-src-modifications]
+
+*See the following sections for relevant unresolved questions:*
+
+- [*Should `rust-src` be a default component?*][unresolved-rust-src]
+
+## Panic strategies
+[panic-strategies]: #panic-strategies
+
+Panic strategies are unlike other profile settings insofar as they influence
```

Comment:
Putting this here since there's not a great place to put it: while it's okay to make a future addition, I think it's good to mention the `optimize_for_size` feature as one of the other main things that people care about when building `std`. There are potential flags we could check for whether to enable this feature, but it's probably going to want its own dedicated flag associated with `build-std` and thus we shouldn't infer it here.

For example, I can imagine `-O2` plus `optimize_for_size` being useful because you're willing to permit, for example, loop unrolling or autovectorization, but not large tables being included in the binary. So, simply checking for `-Os` or `-Oz` is not adequate.